tests/infra_test.go

```diff
@@ -54,6 +55,121 @@ var _ = Describe("Infrastructure", func() {
 	virtClient, err := kubecli.GetKubevirtClient()
 	tests.PanicOnError(err)
 
+	// start a VMI, wait for it to run and return the node it runs on
+	startVMI := func(vmi *v1.VirtualMachineInstance) string {
+		By("Starting a new VirtualMachineInstance")
+		obj, err := virtClient.
+			RestClient().
+			Post().
+			Resource("virtualmachineinstances").
+			Namespace(tests.NamespaceTestDefault).
+			Body(vmi).
+			Do().Get()
+		Expect(err).ToNot(HaveOccurred(), "Should create VMI")
+
+		By("Waiting until the VM is ready")
+		return tests.WaitForSuccessfulVMIStart(obj)
+	}
+
+	Describe("Prometheus scraped metrics", func() {
+
+		/*
+			This test is querying the metrics from Prometheus *after* they were
+			scraped and processed by the different components on the way.
+		*/
+
+		tests.BeforeAll(func() {
+			onOCP, err := clusterutil.IsOnOpenShift(virtClient)
+			Expect(err).ToNot(HaveOccurred(), "failed to detect cluster type")
+
+			if !onOCP {
+				Skip("test is verifying integration with OCP's cluster monitoring stack")
+			}
+		})
+
+		It("should find VMI namespace on namespace label of the metric", func() {
+
+			/*
+				This test is required because in cases of misconfigurations on
+				monitoring objects (such for the ServiceMonitor), our rules will
+				still be picked up by the monitoring-operator, but Prometheus
+				will fail to load it.
+			*/
+
+			By("creating a VMI in a user defined namespace")
+			vmi := tests.NewRandomVMIWithEphemeralDisk(
+				tests.ContainerDiskFor(tests.ContainerDiskAlpine))
+			startVMI(vmi)
+
+			By("finding virt-operator pod")
+			ops, err := virtClient.CoreV1().Pods("kubevirt").
+				List(metav1.ListOptions{LabelSelector: "kubevirt.io=virt-operator"})
+			Expect(err).ToNot(HaveOccurred(), "failed to list virt-operators")
+			Expect(ops.Size).ToNot(Equal(0), "no virt-operators found")
+			op := ops.Items[0]
+			Expect(op).ToNot(BeNil(), "virt-operator pod should not be nil")
+
+			By("finding Prometheus endpoint")
+			ep, err := virtClient.CoreV1().Endpoints("openshift-monitoring").
+				Get("prometheus-k8s", metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred(), "failed to retrieve Prometheus endpoint")
+			promIP := ep.Subsets[0].Addresses[0].IP
+			Expect(promIP).ToNot(Equal(0), "could not get Prometheus IP from endpoint")
+			var promPort int32
+			for _, port := range ep.Subsets[0].Ports {
+				if port.Name == "web" {
+					promPort = port.Port
+				}
+			}
+			Expect(promPort).ToNot(Equal(0), "could not get Prometheus port from endpoint")
+
+			// We need a token from a service account that can view all namespaces in the cluster
+			By("extracting virt-operator sa token")
+			token, _, err := tests.ExecuteCommandOnPodV2(virtClient,
```

Comment:
This looks very sneaky. Isn't there another solution? I.e. can't we just create a test token or something?

Reply:
we can, but it complicates things.. it will require to create a service account just for the test, extract it's token and then tear it down after the test if over... at first, I implemented it that way but then I thought that it will be better to just reuse the operator service account just like we're using the handler pod to query to metrics.